Docs workshop mark maxime (#6725)
RasaHQ/rasa: docs/docs/policies.mdx

In the case that two policies predict with equal confidence (for example, the Memoization
and Rule Policies might both predict with confidence 1), the priority of the
policies is considered. Rasa Open Source policies have default priorities that are set to ensure the
expected outcome in the case of a tie. They look like this, where higher numbers have higher priority:

1. `TEDPolicy`

2. `MemoizationPolicy` and `AugmentedMemoizationPolicy`

> 3. `RulePolicy`

Author: this doesn't render correctly if we don't do this

In general, it is not recommended to have more
than one policy per priority level. If you have 2 policies with the same priority and they predict
with the same confidence, the resulting action will be chosen randomly.

> (TODO: why we do it in number 2. then?)

Author: this is a placeholder: we say that it's a bad idea to have 2 policies with the same priority, but we give it as an example above

Author: oh, yes that's clearer!

In the case that two policies predict with equal confidence (for example, the Memoization
and Rule Policies might both predict with confidence 1), the priority of the
policies is considered. Rasa Open Source policies have default priorities that are set to ensure the
expected outcome in the case of a tie. They look like this, where higher numbers have higher priority:

<!-- We want to have high priority policies first; it's not possible to use a Markdown ordered list for that. -->

* 6 - `RulePolicy`

* 3 - `MemoizationPolicy` or `AugmentedMemoizationPolicy`

* 1 - `TEDPolicy`

In general, it is not recommended to have more

> than one policy per priority level. If you have 2 policies with the same priority and they predict

Reviewer: ```suggestion
than one policy per priority level in your configuration. If you have 2 policies with the same priority and they predict
```

This policy has a pre-defined architecture, which comprises the
following steps:

1. concatenate user input (user intent and entities), previous system actions, slots and active forms for each time
  step into an input vector to pre-transformer embedding layer;

2. feed it to transformer;

3. apply a dense layer to the output of the transformer to get embeddings of a dialogue for each time step;

4. apply a dense layer to create embeddings for system actions for each time step;

5. calculate the similarity between the dialogue embedding and embedded system actions.
   This step is based on the [StarSpace](https://arxiv.org/abs/1709.03856) idea.

It is recommended to use `LabelTokenizerSingleStateFeaturizer`
(see [Featurization of Conversations](./policies.mdx#featurizers) for details).

```yaml-rasa title="config.yml"
policies:
  - name: "TEDPolicy"
    featurizer:
    - name: MaxHistoryTrackerFeaturizer
      max_history: 5
      state_featurizer:
        - name: LabelTokenizerSingleStateFeaturizer

> ```

Author: this is a placeholder: this section didn't contain codeblocks and we thought it'd be a great idea to show the user how to configure the policy. But we didn't know what the content should be, please advise!

Reviewer: showing how to pass the featurizer is great 👍 

on content, @Ghostvv  does it look good?

Reviewer 2: `LabelTokenizerSingleStateFeaturizer` is deprecated, we have only `SingleStateFeaturizer` now

   This step is based on the [StarSpace](https://arxiv.org/abs/1709.03856) idea.

It is recommended to use `LabelTokenizerSingleStateFeaturizer`
(see [Featurization of Conversations](./policies.mdx#featurizers) for details).

```yaml-rasa title="config.yml"
policies:
  - name: "TEDPolicy"
    featurizer:
    - name: MaxHistoryTrackerFeaturizer
      max_history: 5
      state_featurizer:
        - name: LabelTokenizerSingleStateFeaturizer
```

**Configuration:**

You can pass configuration parameters to the `TEDPolicy` using the `config.yml` file.
If you want to fine-tune your model, start by modifying the following parameters:

* `epochs`:
  This parameter sets the number of times the algorithm will see the training data (default: `1`).
  One `epoch` is equals to one forward pass and one backward pass of all the training examples.
  Sometimes the model needs more epochs to properly learn.
  Sometimes more epochs don't influence the performance.
  The lower the number of epochs the faster the model is trained. Here is how the config would look like:

  ```yaml-rasa title="config.yml"
  policies:
  - name: TEDPolicy

>     # if not an appropriate number, it'll use the default value

Reviewer: I don't think this comment is necessary, its not checking for appropriateness (except maybe negative) -- the old comment was basically "pass a number or itll use the default" but also make sure your number is appropriate

This policy has a pre-defined architecture, which comprises the
following steps:

1. Concatenate user input (user intent and entities), previous system actions, slots and active forms for each time
  step into an input vector to pre-transformer embedding layer.

2. Feed the input vector into a transformer.

3. Apply a dense layer to the output of the transformer to get embeddings of a dialogue for each time step.

4. Apply a dense layer to create embeddings for system actions for each time step.

5. Calculate the similarity between the dialogue embedding and embedded system actions.
   This step is based on the [StarSpace](https://arxiv.org/abs/1709.03856) idea.

> It is recommended to use `SingleStateFeaturizer` (see [Featurization](#featurizers) for details).

Author: It looks like we can remove this bit, see [this Slack thread](https://rasa-hq.slack.com/archives/C016FLYV4QN/p1601896968087000) and comment about the `max_history` param. But I don't think I can do this myself as I don't have a clear understanding of the features (pun!) we're talking about 🤷

  One `epoch` is equals to one forward pass and one backward pass of all the training examples.
  Sometimes the model needs more epochs to properly learn.
  Sometimes more epochs don't influence the performance.
  The lower the number of epochs the faster the model is trained.
  Here is how the config would look like:

  ```yaml-rasa title="config.yml"
  policies:
  - name: TEDPolicy
    epochs: 200
  ```

* `max_history`:
  This parameter controls how much dialogue history the model looks at to decide which
  action to take next. Default `max_history` for this policy is `None`,

>   which means that the complete dialogue history since session restart is taken in the

Author: ```suggestion
  which means that the complete dialogue history since session restart is taken into
```

>   account. If you want to limit the model to see only certain number of previous

Author: ```suggestion
  account. If you want to limit the model to only see a certain number of previous
```

  dialogue turns, you can set `max_history` to a finite value.
  Please note that you should pick `max_history` carefully, so that the model has enough
  previous dialogue turns to create a correct prediction.

>   See [Featurizers](#featurizers) for details.

Author: ```suggestion
  See [Featurizers](#featurizers) for more details.
```

* `weight_sparsity`:
  This parameter defines the fraction of kernel weights that are set to 0 for all feed forward layers
  in the model (default: `0.8`). The value should be a number between 0 and 1. If you set `weight_sparsity`
  to 0, no kernel weights will be set to 0, the layer acts as a standard feed forward layer. You should not
  set `weight_sparsity` to 1 as this would result in all kernel weights being 0, i.e. the model is not able
  to learn.

:::tip speeding up training
Default `max_history` for this policy is `None`, which means it'll use the
[`FullDialogueTrackerFeaturizer`](#1-full-dialogue). We recommend to set `max_history` to a finite value in order to
use [`MaxHistoryTrackerFeaturizer`](#2-max-history) for **faster training**.

> See [Featurization of Conversations](./policies.mdx#featurizers) for

Reviewer: ```suggestion
See [Featurizers](./policies.mdx#featurizers) for
```

+---------------------------------+------------------+--------------------------------------------------------------+
| max_relative_position           | None             | Maximum position for relative embeddings.                    |
+---------------------------------+------------------+--------------------------------------------------------------+

> | batch_size                      | [64, 256]        | Initial and final value for batch sizes.                     |

Reviewer: recheck this

with the reduced history. It predicts the next action with confidence `1.0`
if a match is found, otherwise it predicts `None` with confidence `0.0`.

:::note Slots and predictions
If you have dialogues where some slots that are set during
prediction time might not be set in training stories (e.g. in training

> stories starting with a [reminder](./reaching-out-to-user.mdx#reminders) not all previous slots are set),

Reviewer: ```suggestion
stories starting with a [reminder](./reaching-out-to-user.mdx#reminders), not all previous slots are set),
```

### Rule Policy

The `RulePolicy` is a policy that handles conversation parts that follow
a fixed behavior (e.g. business logic). It makes predictions based on
any `rules` you have in your training data. See the
[Rules documentation](./rules.mdx) for further information on how to define rules.

> The `RulePolicy` has the following configuration options:

Author: I unified with above, but I think we never want backticks, right?

Reviewer: really? I think we want to code-ify anything that is code, and i would think the names of policies would be the same. I'm not sure though. @karen-white  what do you think?

Author: the reason I wrote the comment is because I looked at the history of the file and saw [this commit](https://github.com/RasaHQ/rasa/commit/e72ed84c12a2771835eedbe7efba84d3d6036fc3)

Reviewer: I added this to a list of things we need to clarify in the style guide 👍

      - action: utter_greet  # `utter_greet` contradicts `utter_chitchat` from the rule above
    ```
 * `restrict_rules` (default: `true`): Rules are restricted to one user turn, but
    there can be multiple bot events, including e.g. a form being filled and its subsequent submission.

>     Change this parameter to `false` at your own risk.

Reviewer: Can we rephrase this 😅

Author: what about "Changing this parameter to `false` may result in unexpected behaviors."?

Reviewer: Sounds much better 😄

      - action: utter_greet  # `utter_greet` contradicts `utter_chitchat` from the rule above
    ```
 * `restrict_rules` (default: `true`): Rules are restricted to one user turn, but
    there can be multiple bot events, including e.g. a form being filled and its subsequent submission.

>     Changing this parameter to `false` may result in unexpected behaviors.

Reviewer: ```suggestion
    Changing this parameter to `false` may result in unexpected behavior.
```

Every event in a trackers history creates a new state (e.g. running a bot
action, receiving a user message, setting slots). Featurizing a single state
of the tracker has two steps:

> <!-- TODO: below bullet points are not appearing - showing as one block of text instead --->

Author: it looks to me that this is a theme issue. I think @lunelson has this on his TODO

Reviewer 3: Ah ok super!

Author: Lu fixed this in 0.8.0 version of the them. I'll remove the TODO

### Data Augmentation

> When you train a model, by default Rasa Core will create

Reviewer: ```suggestion
When you train a model, by default Rasa Open Source will create
```

Reviewer: we should catch all of these

Every event in a trackers history creates a new state (e.g. running a bot
action, receiving a user message, setting slots). Featurizing a single state
of the tracker has two steps:

1. **Tracker provides a bag of active features**:

    * features indicating intents and entities, if this is the first
     state in a turn, e.g. it's the first action we will take after
     parsing the user's message. (e.g.
     `[intent_restaurant_search, entity_cuisine]` )

    * features indicating which slots are currently defined, e.g.
     `slot_location` if the user previously mentioned the area
     they're searching for restaurants.

    * features indicating the results of any API calls stored in
     slots, e.g. `slot_matches`

    * features indicating what the last bot action or bot utterance was (e.g.
     `prev_action_listen`)

>     * features indicating if any loop is active and which one

Reviewer: Sadly this is not rendering well. I think it is supposed to work with the latest update to the theme -- but i think you want the asterisks to be on the same indentation as the first asterisk above (currently one space in)

Author: ok let me see

Reviewer: just kidding, i think i hadn't installed the latest update 🙂

Author: ahah! then greaaaaat 💯

2. **Convert all the features into numeric vectors**:

    `SingleStateFeaturizer` uses the Rasa NLU pipeline to convert the intent and
    bot action names or bot utterances into numeric vectors.

>     See our [NLU Model Configuration](./model-configuration.mdx) documentation

Reviewer: ```suggestion
    See the [NLU Model Configuration](./model-configuration.mdx) documentation
```
